File: Java/753.java
__________________________________________________________________________________________________
sample 1 ms submission
class Solution {
    private int n;
    private int k;
    private int groupsize;
    private boolean[] used;
    
    public String crackSafe(int n, int k) {
        this.n = n;
        this.k = k;
        
        groupsize = (int) Math.pow(k, n - 1);
        
        int size = (int) Math.pow(k, n);
        used = new boolean[size];
        
        StringBuilder sb = new StringBuilder();
        helper(sb, 0, size);
        
        for (int i = 0; i < n - 1; i++)
            sb.append("0");
        
        return sb.toString();
    }
    
    private boolean helper(StringBuilder sb, int lastNum, int unused) {
        if (unused == 0)
            return true;
        
        int newNumBase = (lastNum % groupsize) * k;
        
        for (int i = 0; i < k; i++)
        {
            int newNum = newNumBase + i;
            if (used[newNum])
                continue;
            
            used[newNum] = true;
            unused--;
            
            if (helper(sb, newNum, unused))
            {
                sb.append(i);
                return true;
            }
            
            used[newNum] = false;
            unused++;
        }
        
        return false;
    }
}
__________________________________________________________________________________________________
sample 36680 kb submission
class Solution {
    public String crackSafe(int n, int k) {
        int M = (int) Math.pow(k, n-1);
        int[] P = new int[M * k];
        for (int i = 0; i < k; ++i)
            for (int q = 0; q < M; ++q)
                P[i*M + q] = q*k + i;

        StringBuilder ans = new StringBuilder();
        for (int i = 0; i < M*k; ++i) {
            int j = i;
            while (P[j] >= 0) {
                ans.append(String.valueOf(j / M));
                int v = P[j];
                P[j] = -1;
                j = v;
            }
        }

        for (int i = 0; i < n-1; ++i)
            ans.append("0");
        return new String(ans);
    }
   
}
__________________________________________________________________________________________________

File: Python3/753.py
__________________________________________________________________________________________________
sample 32 ms submission
class Solution:
    def crackSafe(self, n: int, k: int) -> str:
        M = k**(n-1)
        P = [q*k+i for i in range(k) for q in range(M)]
        ans = []

        for i in range(k**n):
            j = i
            while P[j] >= 0:
                ans.append(str(j // M))
                P[j], j = -1, P[j]

        return "".join(ans) + "0" * (n-1)
__________________________________________________________________________________________________
sample 13388 kb submission
import itertools

class Solution:
    def crackSafe(self, n: int, k: int) -> str:
        def dfs(node, ans):
            while d[node] != 0:
                d[node] -= 1
                dfs(node[1:] + str(d[node]), ans)
            ans.append(node[0])
        if n == 1:
            return ''.join(map(str, range(k)))
        d = {''.join(p): k for p in itertools.product(map(str, range(k)), repeat=n-1)}
        ans = ['0'] * (n-2)
        # dfs('0' * (n-1), ans)
        nodes = ['0' * (n-1)]
        while nodes:
            node = nodes[-1]
            if d[node] != 0:
                d[node] -= 1
                nodes.append(node[1:] + str(d[node]))
                continue
            ans.append(node[0])
            nodes.pop()
        return ''.join(ans)
__________________________________________________________________________________________________
